src/Uri.php

```diff
@@ -254,16 +252,20 @@ private function buildPrimaryIdentifier()
     {
         $identifier = '';
 
-        if ($this->parts['scheme']) {
+        if ($this->parts['scheme'] !== '') {
             $identifier .= $this->parts['scheme'].'://';
         }
 
         $identifier .= $this->authority.$this->parts['path'];
 
-        if ($this->parts['query']) {
+        if ($this->parts['query'] !== '') {
             $identifier .= '?'.$this->parts['query'];
         }
 
+        if ($this->parts['fragment'] !== '' && $this->parts['fragment'][0] !== '/') {
```

Comment:
This isn't right: as per RFC 3986, a fragment denotes a secondary resource; it can't be part of the primary resource identifier.